src/audio/copier/copier.c

```diff
@@ -1716,7 +1716,7 @@ static uint64_t copier_get_processed_data(struct comp_dev *dev, uint32_t stream_
 {
 	struct copier_data *cd = comp_get_drvdata(dev);
 	uint64_t ret = 0;
-	bool source = dev->direction == SOF_IPC_STREAM_CAPTURE;
+	bool source = dev->direction == SOF_IPC_STREAM_PLAYBACK;
```

Comment:
I think you're right, yes. But please also move this variable definition to between lines 1723 and 1724 below - to where it's needed and where it's clear, that it's only used for the host type copier.
TBH, all the `if`s on lines 1721-1724 look suspicious to me, I wouldn't be surprised if they had errors. E.g. IIUC `endpoint_num` is always equal to 1 for DAI, host and gateway copier types and 0 otherwise. The check in line 1721 checks, that it's one of those copier types AND that the direction is correct. Now if it is one of those types but the direction is wrong, we take the `else` path on line 1731, is that correct? In fact that direction test should never fail AFAICS, so if it does fail it should be an error condition?

Reply:
this is part of code from coe, can we use another patch to change it? or you want change it in this patch?

Reply:
> 2\. regarding move 1719 to host part, it is fine for this patch, however, check below, later dai will also have similar implementation, please check below:
>         [a899b9c](https://github.com/thesofproject/sof/commit/a899b9caa61f932368d0936b15c34e89b9553bc8)
>         also need move?

Yes, I'd move it there too. I find it usually better to limit the scope of variables to where they're used

Reply:
@softwarecki could you please double-check that the conditional logic there is correct and that my suspicions in https://github.com/thesofproject/sof/pull/7303/commits/24171f66946486d247dcd60e557cb361a8574e78#r1141840437 are wrong?

Reply:
@lyakh , how about just merge this first and use another PR to track the logic changes?

Reply:
if (cd->endpoint_num && cd->bsource_buffer != input) {
		if (stream_no < cd->endpoint_num) {

change to:
if (cd->endpoint_num)
?

then else part will make more sense.
in dai, I already chagned to switch for this part:

			switch (dev->ipc_config.type) {
			case  SOF_COMP_HOST:
				/* need handle gtw case separately,
				 *since cd->hd not assigned in this case.
				 */
				if (cd->ipc_gtw) {
					struct comp_dev *host_dev;

					host_dev = cd->endpoint[stream_no];
					ret = comp_get_total_data_processed(host_dev,
									    0, input);
				} else if (source == input) {
					ret = cd->hd->total_data_processed;
				}
				break;

			case  SOF_COMP_DAI:
				if (source == input)
					ret = cd->dd->total_data_processed;
				break;

			default:
				ret = comp_get_total_data_processed(cd->endpoint[stream_no],
							    0, input);
				break;

			}